File: composeApp/src/androidMain/kotlin/com/snapchef/app/features/auth/presentation/SignUpViewModel.kt
package com.snapchef.app.features.auth.presentation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.snapchef.app.core.di.SnapChefServiceLocator
import io.ktor.client.plugins.ClientRequestException
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

data class SignUpUiState(
    val name: String = "",
    val email: String = "",
    val password: String = "",
    val showPassword: Boolean = false,
    val agreeTerms: Boolean = false,
    val isLoading: Boolean = false,
    val errorMessage: String? = null
)

class SignUpViewModel : ViewModel() {
    private val _uiState = MutableStateFlow(SignUpUiState())
    val uiState: StateFlow<SignUpUiState> = _uiState.asStateFlow()

    private val apiService = SnapChefServiceLocator.authApiService

    fun updateName(value: String) {
        _uiState.value = _uiState.value.copy(name = value, errorMessage = null)
    }

    fun updateEmail(value: String) {
        _uiState.value = _uiState.value.copy(email = value, errorMessage = null)
    }

    fun updatePassword(value: String) {
        _uiState.value = _uiState.value.copy(password = value, errorMessage = null)
    }

    fun toggleShowPassword() {
        _uiState.value = _uiState.value.copy(showPassword = !_uiState.value.showPassword)
    }

    fun setAgreeTerms(value: Boolean) {
        _uiState.value = _uiState.value.copy(agreeTerms = value)
    }

    fun signUp(onVerifyRequired: (String) -> Unit, onSuccess: () -> Unit) {
        val name = _uiState.value.name.trim()
        val email = _uiState.value.email.trim()
        val password = _uiState.value.password

        if (name.isBlank() || email.isBlank() || password.isBlank()) {
            _uiState.value = _uiState.value.copy(errorMessage = "Please fill out all fields.")
            return
        }

        if (name.length < 2) {
            _uiState.value = _uiState.value.copy(errorMessage = "Name must be at least 2 characters long.")
            return
        }

        val emailRegex = "^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$".toRegex()
        if (!email.matches(emailRegex)) {
            _uiState.value = _uiState.value.copy(errorMessage = "Please enter a valid email address.")
            return
        }

        if (password.length < 8) {
            _uiState.value = _uiState.value.copy(errorMessage = "Password must be at least 8 characters long.")
            return
        }

        if (!_uiState.value.agreeTerms) {
            _uiState.value = _uiState.value.copy(errorMessage = "You must agree to the Terms of Service.")
            return
        }

        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, errorMessage = null)
            try {
                val response = apiService.signup(com.snapchef.app.features.auth.data.remote.SignupRequest(email, name, password))
                onVerifyRequired(response.email)
            } catch (e: ClientRequestException) {
                val status = e.response.status.value
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = when (status) {
                        400, 409 -> "This email is already registered. Please sign in."
                        403      -> { onVerifyRequired(email); return@launch }
                        else     -> "An error occurred during account creation."
                    }
                )
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(
                    isLoading = false,
                    errorMessage = "Network error. Please check your connection."
                )
            }
        }
    }
}

File: iosApp/iosApp/auth_screens/SignUpViewModel.swift
//
//  SignUpViewModel.swift
//  iosApp
//
//  Created by gergana on 3/28/26.
//

import Foundation
import Shared

@MainActor
final class SignUpViewModel: ObservableObject {

    @Published var name:          String  = ""
    @Published var email:         String  = ""
    @Published var password:      String  = ""
    @Published var showPassword:  Bool    = false
    @Published var agreeTerms:    Bool    = false
    @Published var isLoading:     Bool    = false
    @Published var errorMessage:  String? = nil

    private let apiService = SnapChefServiceLocator.shared.authApiService

    func updateName(_ value: String) {
        name         = value
        errorMessage = nil
    }

    func updateEmail(_ value: String) {
        email        = value
        errorMessage = nil
    }

    func updatePassword(_ value: String) {
        password     = value
        errorMessage = nil
    }

    func toggleShowPassword() {
        showPassword.toggle()
    }

    func setAgreeTerms(_ value: Bool) {
        agreeTerms = value
    }

    func signUp(
        onVerifyRequired: @escaping (String) -> Void,
        onSuccess:        @escaping () -> Void
    ) {
        let trimmedName  = name.trimmingCharacters(in: .whitespaces)
        let trimmedEmail = email.trimmingCharacters(in: .whitespaces)

        guard !trimmedName.isEmpty, !trimmedEmail.isEmpty, !password.isEmpty else {
            errorMessage = "Please fill out all fields."
            return
        }

        guard trimmedName.count >= 2 else {
            errorMessage = "Name must be at least 2 characters long."
            return
        }

        let emailRegex = #"^[A-Za-z0-9+_.-]+@[A-Za-z0-9.-]+$"#
        guard trimmedEmail.range(of: emailRegex, options: .regularExpression) != nil else {
            errorMessage = "Please enter a valid email address."
            return
        }

        guard password.count >= 8 else {
            errorMessage = "Password must be at least 8 characters long."
            return
        }

        guard agreeTerms else {
            errorMessage = "You must agree to the Terms of Service."
            return
        }

        isLoading    = true
        errorMessage = nil

        Task {
            do {
                let response = try await apiService.signup(
                    request: SignupRequest(email: trimmedEmail, name: trimmedName, password: password)
                )
                onVerifyRequired(response.email)
            } catch let error as KotlinException {
                isLoading = false
                let code  = httpStatusCode(from: error)
                switch code {
                case 400, 409:
                    errorMessage = "This email is already registered. Please sign in."
                case 403:
                    onVerifyRequired(trimmedEmail)
                default:
                    errorMessage = "An error occurred during account creation."
                }
            } catch {
                isLoading    = false
                errorMessage = "Network error. Please check your connection."
            }
        }
    }

    func googleSignIn(
        idToken:   String,
        onSuccess: @escaping () -> Void
    ) {
        isLoading    = true
        errorMessage = nil

        Task {
            do {
                let response = try await apiService.googleAuth(idToken: idToken)
                AuthManager.shared.accessToken = response.accessToken
                AuthManager.shared.currentUser = response.user
                onSuccess()
            } catch {
                isLoading    = false
                errorMessage = "Google Sign-In failed on the server."
            }
        }
    }

    private func httpStatusCode(from error: KotlinException) -> Int {
        let desc = error.description()
        let numbers = desc.components(separatedBy: CharacterSet.decimalDigits.inverted)
        return numbers.compactMap(Int.init).first(where: { $0 >= 400 && $0 < 600 }) ?? -1
    }
}
